### PR TITLE
[YAML] Add _ById commands support for darwin-framework-tool

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

This PR introduces partial support for *ById YAML commands in the `darwin-framework-tool` backend.
By **partial** I mean that while basic tests can be written, the complete tests (such as **TestAttributesById.yaml** and **TestCommandsById.yaml**) are not fully supported.
This is because `darwin-framework-tool` currently lacks features like handling multiple attributes, as well as the ability to return  `UNSUPPORTED_ENDPOINT` when targeting a nonexistent endpoint.

For now, this is mainly useful for testing specific code paths under certain scenarios that uses those commands.

### CODE REVIEW

File: examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py [hunk 1]
```diff
@@ -348,7 +385,7 @@ def run(self, specs, value, cluster_name: str, typename: str, array: bool):
                 del value[key_name]
 
         elif isinstance(value, list) and array:
-            value = [self.run(specs, v, cluster_name, typename, False)
+            value = [self.run(specs, v, cluster_name, typename, isinstance(v, list))
```
Comment (bzbarsky-apple) [inline examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py:388]: Hmm, why is this needed?  Are there lists of lists in the spec somewhere?

File: examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py [hunk 2]
```diff
@@ -252,6 +253,42 @@ def maybe_convert(self, typename: str, value):
...
+            value = data.get('value')
+            value_type = data.get('type')
+
+            if isinstance(value, list):
+                if value_type == 'Structure':
+                    struct = {}
+                    for field in value:
+                        context_tag = field.get('contextTag')
+                        struct[str(context_tag)] = self.__convert_data_to_value(
+                            specs, field, cluster_name, typename)
```
Comment (bzbarsky-apple) [inline examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py:283]: Is using `typename` here really the right thing?  Or should it be the name of the type of the field?

The fact that this is very unclear suggests this method needs more docs on the arguments...

File: examples/darwin-framework-tool/commands/clusters/ClusterCommandBridge.h [hunk 3]
```diff
@@ -73,10 +73,14 @@ class ClusterCommand : public ModelCommand {
         uint16_t __block responsesNeeded = repeatCount;
         dispatch_queue_t callbackQueue = dispatch_queue_create("com.chip.command", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
 
+        __block auto * endpoint = [NSNumber numberWithUnsignedShort:endpointId];
+        __block auto * cluster = [NSNumber numberWithUnsignedInteger:clusterId];
+        __block auto * command = [NSNumber numberWithUnsignedInteger:commandId];
```
Comment (bzbarsky-apple) [inline examples/darwin-framework-tool/commands/clusters/ClusterCommandBridge.h:78]: Why do these need to be `__block`?  They don't get mutated, right?

Also, any reason to not use `@(endpointId)` etc to box?

File: examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h [hunk 4]
```diff
@@ -57,20 +57,26 @@ class ReadAttribute : public ModelCommand {
         if (mFabricFiltered.HasValue()) {
             params.filterByFabric = mFabricFiltered.Value();
         }
+
+        __block auto * endpoint = [NSNumber numberWithUnsignedShort:endpointId];
+        __block auto * cluster = [NSNumber numberWithUnsignedInteger:mClusterId];
+        __block auto * attribute = [NSNumber numberWithUnsignedInteger:mAttributeId];
```
Comment (bzbarsky-apple) [inline examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:63]: Same questions about `__block` and boxing here.

File: examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h [hunk 5]
```diff
@@ -137,16 +143,23 @@ class SubscribeAttribute : public ModelCommand {
             params.resubscribeAutomatically = mAutoResubscribe.Value();
         }
 
-        [device subscribeToAttributesWithEndpointID:[NSNumber numberWithUnsignedShort:endpointId]
-            clusterID:[NSNumber numberWithUnsignedInteger:mClusterId]
-            attributeID:[NSNumber numberWithUnsignedInteger:mAttributeId]
+        __block auto * endpoint = [NSNumber numberWithUnsignedShort:endpointId];
+        __block auto * cluster = [NSNumber numberWithUnsignedInteger:mClusterId];
+        __block auto * attribute = [NSNumber numberWithUnsignedInteger:mAttributeId];
```
Comment (bzbarsky-apple) [inline examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:148]: As above.

File: examples/darwin-framework-tool/commands/clusters/WriteAttributeCommandBridge.h [hunk 6]
```diff
@@ -74,10 +74,13 @@ class WriteAttribute : public ModelCommand {
     {
         dispatch_queue_t callbackQueue = dispatch_queue_create("com.chip.command", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
 
+        __block auto * endpoint = [NSNumber numberWithUnsignedShort:endpointId];
+        __block auto * cluster = [NSNumber numberWithUnsignedInteger:mClusterId];
+        __block auto * attribute = [NSNumber numberWithUnsignedInteger:mAttributeId];
```
Comment (bzbarsky-apple) [inline examples/darwin-framework-tool/commands/clusters/WriteAttributeCommandBridge.h:79]: As above.

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py (hunk 1 above)
Comment (vivien-apple) [reply examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py:388]: > Hmm, why is this needed? Are there lists of lists in the spec somewhere?

Not really. This is mainly due to the return format used for API return values such as [1].

For instance, if you run the following command:

```
$ darwin-framework-tool any read-by-id 29 0 0x12344321 65535
```

The return value will be:

```
[[{'DeviceType': 17, 'Revision': 1}, {'DeviceType': 22, 'Revision': 1}], [{'DeviceType': 256, 'Revision': 1}, {'DeviceType': 17, 'Revision': 1}], [{'DeviceType': 256, 'Revision': 1}, {'DeviceType': 17, 'Revision': 1}], [{'DeviceType': 15, 'Revision': 3}], [{'DeviceType': 15, 'Revision': 3}]]
```

This format returns a list of lists containing dictionaries.

[1] https://github.com/project-chip/connectedhomeip/blob/master/examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h#L60

File: examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py (hunk 2 above)
Comment (vivien-apple) [reply examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py]: Apologies for that!
I was a bit overzealous with the arguments. We could simplify it by passing a single argument instead of four, which should make it clearer.

File: examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py:389]: Ah, OK.  Might be worth documenting that.

**Review by Sum1cares — APPROVED** [review]
